internal/migration/migration.go

```diff
@@ -0,0 +1,478 @@
+package migration
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	json "github.com/nikkolasg/hexjson"
+	"go.etcd.io/bbolt"
+
+	"github.com/drand/drand/chain"
+	"github.com/drand/drand/chain/postgresdb/database"
+	"github.com/drand/drand/chain/postgresdb/pgdb"
+	"github.com/drand/drand/chain/postgresdb/schema"
+	"github.com/drand/drand/log"
+)
+
+type (
+	beacon struct {
+		PreviousSig []byte `json:",omitempty"`
+		Round       uint64
+		Signature   []byte
+	}
+
+	migrator struct {
+		ctx       context.Context
+		ctxCancel context.CancelFunc
+
+		startedAt time.Time
+		logger    log.Logger
+
+		pgDSN            string
+		bufferSize       int
+		beaconName       string
+		sourceBeaconPath string
+
+		// existingDB   *bbolt.DB
+		existingRows int
+
+		distChan    chan beacon
+		errChan     chan error
+		destination chain.StorageType
+	}
+)
+
+const ownerOnly = 0600
+
+const DefaultBufferSize = 10_000
+
+var (
+	// ErrMigrationNotNeeded ...
+	ErrMigrationNotNeeded = fmt.Errorf("migration not needed")
+
+	bucketName = []byte("beacons")
+)
+
+//nolint:lll // This function uses the right amount of params
+func Migrate(ctx context.Context, logger log.Logger, sourceBeaconPath, beaconName string, destination chain.StorageType, pgDSN string, bufferSize int) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	startedAt := time.Now()
+
+	if err := shouldMigrate(ctx, logger, sourceBeaconPath, beaconName, destination, pgDSN); err != nil {
+		logger.Warnw("decided storage format migration is not needed", "err", err)
+		if errors.Is(err, ErrMigrationNotNeeded) {
+			return err
+		}
+
+		return nil
+	}
+
+	bufferSize, err := computerBufferSize(bufferSize, logger, sourceBeaconPath)
+	if err != nil {
+		return err
+	}
+
+	m := migrator{
+		logger:    logger,
+		startedAt: startedAt,
+
+		pgDSN:            pgDSN,
+		bufferSize:       bufferSize,
+		beaconName:       beaconName,
+		sourceBeaconPath: sourceBeaconPath,
+		destination:      destination,
+
+		ctx:       ctx,
+		ctxCancel: cancel,
+
+		distChan: make(chan beacon, bufferSize),
+
+		errChan: make(chan error),
+	}
+
+	return m.doMigrate(ctx)
+}
+
+func computerBufferSize(bufferSize int, logger log.Logger, sourceBeaconPath string) (int, error) {
+	switch {
+	case bufferSize < 0:
+		logger.Infow("buffer size not specified, defaulting to 10000")
+		bufferSize = DefaultBufferSize
+	case bufferSize == 0:
+		var err error
+		bufferSize, err = automaticBufferSize(logger, sourceBeaconPath)
+		if err != nil {
+			return 0, err
+		}
+	case bufferSize < DefaultBufferSize:
```

Comment:
Less than 0 will always beat < DefaultBufferSize, so move this to the top.
Let's define the constant in here if this is the only place it is used.

Reply:
I've simplified this and replaced with a few if statements